### DISCUSSION TITLE
receiving same response for different situations

### DISCUSSION THREAD

**Discussion by nikhil-varma-mudunuri** [pocketbase/pocketbase#2944] [Q&A] [answered]
in my app I am trying to implement authentication with pocket base, my app login is login user if the user already exists if not create a user and then login. but receiving the same response from the pocket base for both the userNotFound case and the password incorrect case. Is there any solution?

**Comment by ganigeorgiev** [comment]
The auth errors are deliberately general to prevent _to some extend_ users enumeration (_see for example [OWASP guidelines](https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html#authentication-and-error-messages)_).

There is no way to change the auth response without attaching to the [PocketBase event hooks](https://pocketbase.io/docs/event-hooks/) (for example with the `OnRecordBeforeAuthWithPasswordRequest`; `e.Record` will be nil if no valid auth record with the provided `identity` was found).

Alternatively, depending on your use case you can also try to create the user after a failed login attempt behind the scenes (_this should work as long as the client request satisfy the Create API rule of your auth collection_), something like:

```js
const email = "..."
const pass = "..."

try {
    await pb.collection("users").authWithPassword(email, pass)
} catch (err) {
    await pb.collection("users").create({
        email:           email,
        password:        pass,
        passwordConfirm: pass,
    })
    await pb.collection("users").authWithPassword(email, pass)
}
```

**Reply by nikhil-varma-mudunuri** [reply]
`app.OnRecordBeforeAuthWithPasswordRequest().Add(func(e *core.RecordAuthWithPasswordEvent) error {

		if e.Record == nil {

			collection, err := app.Dao().FindCollectionByNameOrId("USERS")
			if err != nil {
				return err
			}

			record := models.NewRecord(collection)

			form := forms.NewRecordUpsert(app, record)

			form.LoadData(map[string]any{
				"username":        e.Identity,
				"name":            e.Identity,
				"password":        e.Password,
				"passwordConfirm": e.Password,
			})

			if err := form.Submit(); err != nil {
				return err
			}
		}
		return nil
	})`

I am using this code to create a new user if none , I it's not waiting until user record get created is there any way to wait until that and authenticate newly created user

**Reply by ganigeorgiev** [reply, marked as answer]
I'm not sure that I understand what do you mean by _" it's not waiting until user record get created "_. The hooks execute as part of the same goroutine, aka. synchronously/blocking.

Note that in the above you are not changing `e.Record` so the hook will still complete without an auth record. Or in other words, try to replace `record := models.NewRecord(collection)` with `e.Record = models.NewRecord(collection)` and pass `e.Record` to the upsert form (or assign `e.Record = record` after the `Submit()` call).
